sslh-select.c

```diff
@@ -180,6 +182,93 @@ void shovel(struct connection *cnx, int active_fd,
     }
 }
 
+/* shovels data from one fd to the other and vice-versa
+   returns after one socket closed
+ */
+void shovel_single(struct connection *cnx)
+{
+   fd_set fds_r, fds_w;
+   int res, i;
+   int max_fd = MAX(cnx->q[0].fd, cnx->q[1].fd) + 1;
+
+   FD_ZERO(&fds_r);
+   FD_ZERO(&fds_w);
+   while (1) {
+      for (i = 0; i < 2; i++) {
```

Comment:
See below -- deferring is never going to happen in a forked process except for the very first iteration (basically the data that was used for probing the protocol, which is stored in the deferred buffer)

Reply:
Deferring can happen even during shovelling as long as the socket stays nonblocking. I originally forgot that it was nonblocking, copied the shovel function from `sslh-fork` and kept getting SSH connections dropped due to checksum mismatches as soon as the server tried to send a lot of data at once for any reason. I concluded that it was deferring parts of the data and never flushing it out afterwards.

I suppose we could switch the socket to blocking mode after forking, but nonblocking might actually be faster or more reliable? I’m not sure. I even considered that maybe `sslh-fork` should use nonblocking sockets, too. Is it ever possible that blocking shovelling might deadlock? …I guess not. Blocking is probably fine then.